### ISSUE TITLE
How to translate strings with interpolation

### ISSUE THREAD

**Issue by iszak** [alexanderwallin/lioness#10]
Given we have the higher order function `withTranslators` which exposes the translation functions e.g. `tcpn` if I want to translate a string that may not render as HTML (e.g. `document.title`) `tcpn` will interpolate with `<span />`'s which is invalid in the given example.

I suggest a few proposals;
- By default the `tc`, `tcn`, `tcp`, `tcpn` don't bind the interpolator but instead it's done in the `<T />` component as this is likely the use case. If there comes a case where a user wants to use those functions but also interpolate with the span they can import interpolateComponents (perhaps we could add this as a default export to facilitate this more) and pass it as the first parameter to those functions since they will be unbound.

- Alternative is to expose the gettext instance via the context to allow users to do their own interpolation, I am not in favour of this as it adds yet another property to the context and provides a leaky implementation since in all circumstances currently gettext instance is hidden.

In either case, it's probably worth while to add a interpolate function which provides a span-less implementation and export it because both of these solutions will require it, otherwise the end user will have to re-implement it.

Thoughts?

**Comment by iszak** [comment]
After looking into it a bit more, it seems if we do expose two interpolate functions (one for components and one without) they may have different semantics, for example do we expect the one not for components to also support the syntax `{{ element:string }}` where element is a React element? If we do support it, is it reasonable to expect the user to understand the implications?

**Comment by alexanderwallin** [comment]
My first thought was that there are many string interpolation libraries out there. At the same time, it might be inconsistent to leave one single unit out of the package.

What would `{{ element:string }}` do in a string-only case?

Is think `<T>` should always interpolate components, as it is a component itself. We could provide an `interpolate()` function in the context, which itself could just be another library passed on.

**Comment by iszak** [comment]
I agree there are many string interpolation libraries and I think this is why by exposing unbound translation functions through the context is the way to go and instead doing the binding on a `<T />` level. I think as well providing an provider option to configure the interpolation in the context which `<T />` can use is also another good idea to allow people to use interpolation libraries of their own.

However I think we should provide something convenient out of the box which should be the default for `<T />` unless it becomes too cumbersome to parse/support in which case we need to consider;

I think both "string-only" interpolation and `<T />` interpolation should do react element replacements e.g. `{{ element:string }}`, I think where they differ should be that the text-only version shouldn't wrap variable replacements in `<span>`.

Suggested implementation;
- Expose interpolate as configurable via `<Provider />` and expose it in context for `<T />` to use
- Expose generic/string-only interpolate which doesn't wrap variables in `<span />`

**Comment by alexanderwallin** [comment]
> I agree there are many string interpolation libraries and I think this is why by exposing unbound translation functions through the context is the way to go and instead doing the binding in the <Provider /> level. I think as well providing an provider option to configure the interpolation in the context which <T /> can use is also another good idea to allow people to use interpolation libraries of their own.
> 
> However I think we should provide something convenient out of the box which should be the default for <T /> unless it becomes too cumbersome to parse/support in which case we need to consider;

Good points! I agree. This would essentially be two functions that both have the signature `interpolate(str, scope)` then?

> I think both "string-only" interpolation and <T /> interpolation should do react element replacements e.g. {{ element:string }}, I think where they differ should be that the text-only version shouldn't wrap variable replacements in <span>.
> 
> Suggested implementation;
> 
> * Expose interpolate as configurable via <Provider /> and expose it in context for <T /> to use
> * Expose generic/string-only interpolate which doesn't wrap variables in <span />

How would you accomplish this? How do you wrap React components inside a string?

**Comment by iszak** [comment]
I think the signature is good 👍  

For the string-only, I don't think you would wrap it in a react component, rather you compile any react-element replacements to a HTML string and interpolate that in your string and return that - I'm not sure if I'm made this clear.

**Comment by alexanderwallin** [comment]
Right. I think that is working against React and its API; rendering HTML requires you to pass it in the `dangerouslySetInnerHTML` prop.

**Comment by iszak** [comment]
I don't think I was clear enough, the string-only implementation would work as the current but for each variable replacement instead of returning a react element it would return a string which renders the React components to HTML which doesn't require usage of `dangerouslySetInnerHTML` but instead `ReactDOM.render` I can probably get a working example out in the next few days if it's easier to visualise with some code

**Comment by alexanderwallin** [comment]
Oh, okay. Technically I understand. When would you use this?

**Comment by iszak** [comment]
I personally don't have a use-case for the string-only implementation with variable replacement of a stringified react element - it was only suggested to keep consistency between both interpolations. 

My use-case is I want to translate a string with variable replacement, but the current implementation does variable replacements by wrapping it in `<span />` which is invalid in my use case e.g. document.title

**Comment by iszak** [comment]
@alexanderwallin I've pushed an initial implementation on the branch patch-6 https://github.com/alexanderwallin/lioness/compare/master...iszak:patch-6?expand=1 not sure if it's worth creating a pull request, I expect to get through a few iterations

**Comment by alexanderwallin** [comment]
I suggest we only implement solutions to problems we are actually facing. It's a good idea to be consistent, but I feel the added complexity and technical debt is not worth it.

**Comment by iszak** [comment]
Those changes do address a real problem I am facing. I agree it does add additional complexity - as for technical debt, what technical debt is introduced? I am happy to pay down the technical debt if possible.

**Comment by alexanderwallin** [comment]
> Those changes do address a real problem I am facing.

You said above that you personally didn't have a use case? 😊

> I agree it does add additional complexity - as for technical debt, what technical debt is introduced? I am happy to pay down the technical debt if possible.

I guess I just meant the technical implementation and maintenance of the extra complexity.

**Comment by iszak** [comment]
So I don't have a use case for string-only implementation with variable replacement *of a stringified react element* but I do have a use case for string-only implementation with variable replacement without span's

**Comment by alexanderwallin** [comment]
Ah, yes yes. Let's support the latter and ditch the former! I think the plan you outlined [here](https://github.com/alexanderwallin/lioness/issues/10#issuecomment-306525413) is well and good:

> I agree there are many string interpolation libraries and I think this is why by exposing unbound translation functions through the context is the way to go and instead doing the binding in the level. I think as well providing an provider option to configure the interpolation in the context which can use is also another good idea to allow people to use interpolation libraries of their own.
> 
> However I think we should provide something convenient out of the box which should be the default for unless it becomes too cumbersome to parse/support in which case we need to consider;

I can sketch up a quick PR this afternoon.

**Comment by andrewpeterprifer** [comment]
@alexanderwallin is there an update on this? I also have a use case where I'd just need the vanilla string and not a React component.
I think the use case for using the t*() functions directly is pretty much only when you need a string and not a component, so it'd be really cool to change this. :)

**Comment by alexanderwallin** [comment]
How's that looking, @iszak and @andrewpeterprifer?
